contrib/coredns/policy/policy.go

```diff
@@ -149,7 +178,29 @@ func (p *PolicyMiddleware) ServeDNS(ctx context.Context, w dns.ResponseWriter, r
 	}
 
 	if response.Permit {
-		return middleware.NextOrFailure(p.Name(), p.Next, ctx, w, r)
+		status,err := middleware.NextOrFailure(p.Name(), p.Next, ctx, w, r)
```

Comment:
I have implemented structure like this


type LocalResponseWriter struct {
	dns.ResponseWriter
	Rcode int
	Len int
	Msg   *dns.Msg
	localAddr  net.Addr
	remoteAddr net.Addr
}

func New(w dns.ResponseWriter) *LocalResponseWriter {

	var addr string = ""
	l := &net.IPAddr{IP: net.ParseIP(addr)}
	r := &net.IPAddr{IP: net.ParseIP(addr)}
	return &LocalResponseWriter{
		ResponseWriter: w,
		Rcode:          0,
		Msg:            nil,
		localAddr: l,
		remoteAddr: r,

	}
}

func (r *LocalResponseWriter) WriteMsg(res *dns.Msg) error {
	r.Rcode = res.Rcode
	// We may get called multiple times (axfr for instance).
	// Save the last message, but add the sizes.
	r.Len += res.Len()
	r.Msg = res
	return r.ResponseWriter.WriteMsg(res)
}

// Write is a wrapper that records the length of the message that gets written.
func (r *LocalResponseWriter) Write(buf []byte) (int, error) {
	n, err := r.ResponseWriter.Write(buf)
	if err == nil {
		r.Len += n
	}
	return n, err
}

my understanding is that async WriteMsg call will update the message, 

I will make the call with the new structure

localResponseWriter := new(LocalResponseWriter)
		localResponseWriter.localAddr = w.LocalAddr()
		localResponseWriter.remoteAddr = w.RemoteAddr()
		localResponseWriter.Msg = r

		upstreamResult := new(dns.Msg)

		status,err := middleware.NextOrFailure(p.Name(), p.Next, ctx, localResponseWriter, upstreamResult)
		if err != nil {
			return status, err
}

and pick the response

state := request.Request{W: localResponseWriter, Req: upstreamResult}

		rr := upstreamResult.Answer[0]
		switch state.Family() {
		case 1:
			attrs = append(attrs, &pb.Attribute{Id: "address", Type: "address", Value: rr.(*dns.A).A.String()})
		case 2:
			attrs = append(attrs, &pb.Attribute{Id: "address", Type: "address", Value: rr.(*dns.AAAA).AAAA.String()})

		}


I will submit Push Request with changes and will update code if my understanding is not correct.